docs/kuneiform/procedures.md

```diff
@@ -0,0 +1,275 @@
+        // something else
+    }
+}
+```
+
+### For Loop
+
+For loops can be declared over 4 different targets: arrays, ranges, procedure calls, and SQL statements.
+
+#### Array Loop
+
+```ts
+procedure iterate_over_array($names text[]) public {
+    for $name in $names {
+        // $name is of type `text`
+    }
+}
+```
+
+#### Range Loop
+
+```ts
+procedure iterate_over_range() public {
+    // iterating over all numbers from 1 to 25
+    for $i in 1..25 {
+        // $i is type int
+    }
+}
+
+procedure iterate_range_dynamic($start int, $end int) public {
+    // can also iterate over dynamic ranges
+    for $i in $start..$end {
+        // ...
+    }
+}
+```
+
+#### Procedure Loops
+
+You can iterate over any SQL statement as well. The following examples reference an example table `users`, with columns `id`, `name`, and `age`.
+
+```ts
+procedure iterate_over_sql() public {
+    for $row in SELECT id, name, age from users {
+        // can access $row.id, $row.name, and $row.age
+    }
+}
+```
+
+You can also iterate of returning clauses:
+
+```ts
+procedure iterate_over_delete() public {
+    $age int := 100;
+    for $row in delete from users
+        where age = $age
+        returning * {
+        // can access $row.id, $row.name, and $row.age
+        // for all deleted users
+    }
+}
+```
+
+Procedures that return tables can also be iterated over:
+
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    // ...
+}
+
+procedure iterate_over_users() public {
+    for $row in SELECT * FROM get_users() {
+        // can access $row.id, $row.name, and $row.age
+    }
+}
+```
+
+#### Breaking From a Loop
+
+To conditionally exit a loop, you can use `break`:
+
+```ts
+procedure get_users() public {
+    $count int := 0;
+    for $row in SELECT * FROM users {
+        // do something
+
+        // break if too many records have been read
+        $count := $count + 1;
+        if $count >= 50 {
+            break
+        }
+    }
+}
+```
+
+
+
+### Returning Values
+
+#### Return Single Records
+
+To return a single record, simply use the `return` key word:
+
+```ts
+procedure get_numbers() public view returns (num1 int, num2 int) {
+    $number1 int := 1;
+    return $number1, 2;
+}
+```
+
+To return a single row from a SQL query, you should iterate over the result and return on the first iteration:
+
+```ts
+procedure get_user($id uuid) public view returns (name text, age int) {
+    for $row in SELECT name, age from users where id = $id {
+        // this will always return on the first iteration.
+        return $row.name, $row.age;
+    }
+
+    // if no iterations, then there is no matching user.
+    error('user not found');
+}
+```
+
+#### Return Table
+
+To return a table from a procedure, you can simply return any SQL query that returns data:
+
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    return select * from users;
+}
+```
+
+#### Return Next
```

Comment:
Does `return next` return the following iteration in the loop? Because your example makes it seem so.